src/uct/api/uct.h

```diff
@@ -2065,33 +2066,84 @@ ucs_status_t uct_md_query(uct_md_h md, uct_md_attr_t *md_attr);
+ * The enumeration allows specifying which fields in @ref uct_mem_alloc_params_t
+ * are present.
  */
-ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *name, uct_mem_h *memh_p);
+typedef enum {
+    /** Enables @ref uct_mem_alloc_params_t::flags */
+    UCT_MEM_ALLOC_PARAM_FIELD_FLAGS          = UCS_BIT(0),
+
+    /** Enables @ref uct_mem_alloc_params_t::address */
+    UCT_MEM_ALLOC_PARAM_FIELD_ADDRESS        = UCS_BIT(1),
+
+    /** Enables @ref uct_mem_alloc_params_t::mem_type */
+    UCT_MEM_ALLOC_PARAM_FIELD_MEM_TYPE       = UCS_BIT(2),
+
+    /** Enables @ref uct_mem_alloc_params_t::mds */
+    UCT_MEM_ALLOC_PARAM_FIELD_MDS            = UCS_BIT(3),
+
+    /** Enables @ref uct_mem_alloc_params_t::name */
+    UCT_MEM_ALLOC_PARAM_FIELD_NAME           = UCS_BIT(4)
+} uct_mem_alloc_params_field_t;
 
 
 /**
  * @ingroup UCT_MD
- * @brief Release memory allocated by @ref uct_md_mem_alloc.
- *
- * @param [in]     md          Memory domain memory was allocated on.
- * @param [in]     memh        Memory handle, as returned from @ref uct_md_mem_alloc.
+ * @brief Parameters for allocating memory using @ref uct_mem_alloc
  */
-ucs_status_t uct_md_mem_free(uct_md_h md, uct_mem_h memh);
+typedef struct {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_mem_alloc_params_field_t. Fields not specified in this mask will
+     * be ignored
+     */
+    uint64_t                     field_mask;
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned                     flags;
+
+    /**
+     * If @a address is NULL, the underlying allocation routine will
+     * choose the address at which to create the mapping. If @a address
+     * is non-NULL but UCT_MD_MEM_FLAG_FIXED is not set, the address
+     * will be interpreted as a hint as to where to establish the mapping. If
+     * @a address is non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then the
+     * specified address is interpreted as a requirement. In this case, if the
+     * mapping to the exact address cannot be made, the allocation request
+     * fails.
+     */
+    void                         *address;
+
+    /**
+     * Type of memory to be allocated.
+     */
+    ucs_memory_type_t            mem_type;
+
+    struct {
+        /**
+         * Array of memory domains to attempt to allocate
+         * the memory with, for MD allocation method.
+         */
+        const uct_md_h           *mds;
+
+        /**
+         *  Length of 'mds' array. May be empty, in such case
+         *  'mds' may be NULL, and MD allocation method will
+         *  be skipped.
+         */
+        unsigned                 count;
+    } mds;
+
+    /**
+     * Name of the allocated region, used to track memory
+     * usage for debugging and profiling.
+     */
+    const char                   *name;
```

Comment:
@yosefe also, I think we have to define default name ?

Reply:
L2112: too many negatives (non-NULL ... but ... not set).  but -> and?